### PR TITLE
[REVIEW] Implementation of gdf_column_concatenate

### PR DESCRIPTION
**Pull request by harrism**
This implements concatenation of multiple GDF columns, including correct concatenation of validity bit masks, using Thrust and cudaMemcopies.

Note that this PR supersedes @jrhemstad's PR #148 (Jake stubbed it out and added error checking and I took over implementation in my own fork).

Note also that I needed to merge from Jake's branch for null mask counting (#147) into my branch because I wanted the validity mask concatenation to be in the new validops.cu file that Jake created. I hope this merging doesn't cause issues merging both of our PRs. At this time I believe my branch is up to date with his changes.

I have included unit tests and Jake helped suggest some cases to test (and provided some awesome testing utilities).

This is basically ready for review, but I want @dantegd to test it before I switch to [REVIEW].

### CODE REVIEW

**Comment by scopatz** [comment]
Seems like tests timed out.

```
08:05:20 Running tests...
08:05:20 Test project /home/jenkins/workspace/libgdf-prb/build
08:05:20       Start  1: foo-sample-test
08:05:20  1/11 Test  #1: foo-sample-test ..................   Passed    0.00 sec
08:05:20       Start  2: datetimeops_test
08:05:21  2/11 Test  #2: datetimeops_test .................   Passed    1.14 sec
08:05:21       Start  3: hash_test
08:05:29  3/11 Test  #3: hash_test ........................   Passed    8.08 sec
08:05:29       Start  4: join_test
08:10:29 Build timed out (after 5 minutes). Marking the build as aborted.
08:10:29 Build was aborted
08:10:29 Recording test results
08:10:30 Adding one-line test results to commit status...
```

**Comment by jrhemstad** [comment]
rerun tests

**Comment by scopatz** [comment]
Seems to work now

File: src/column.cpp [hunk 1]
```diff
@@ -1,13 +1,171 @@
...
+  //    a) use cudaMemcpyAsync to copy the data and overlap copies with gdf_mask_concat
+  //       (this will require getting rid of the allocations below because they will
+  //       implicitly sync the device), or
+  //    b) use a kernel to copy the data from all columns in one go. This will likely not
+  //       overlap with the gdf_mask_concat
+  // 2. Detect a zero total null count and skip gdf_mask_concat -- use cudaMemsetAsync
+
+  // copy data
+  for (int i = 0; i < num_columns; ++i) {   
+    gdf_size_type bytes = sizeof(column_type) * columns_to_concat[i]->size;
```
[the remark this answers is not in the document]
Comment (harrism) [reply src/column.cpp]: That's a brain fart! Sorry. Why is this now closed?

File: src/column.cpp [hunk 2]
```diff
@@ -1,13 +1,171 @@
...
+    if ((current_column->size > 0) && (nullptr == current_column->data))
+    {
+      return GDF_DATASET_EMPTY;
+    }
+    if (column_type != current_column->dtype) {
+      return GDF_DTYPE_MISMATCH;
+    }
+
+    total_size += current_column->size;
+    at_least_one_mask_present |= (columns_to_concat[i]->valid != 0);
```
[the remark this answers is not in the document]
Comment (harrism) [reply src/column.cpp]: When I started programming C++ we didn't have nullptr! We HAD to use zero or NULL!

http://dilbert.com/strip/2011-06-12

[the remark this answers is not in the document]
Comment (harrism) [reply src/column.cpp]: That's not the dilbert I meant to paste! The one I meant is MUCH older: 
http://dilbert.com/strip/1992-09-08

File: src/column.cpp [hunk 3]
```diff
@@ -1,13 +1,171 @@
...
+    cudaMemcpy(target, columns_to_concat[i]->data, bytes, cudaMemcpyDeviceToDevice);
+    target += bytes;
+
+    output_column->null_count += columns_to_concat[i]->null_count;
+  }
+  
+  if (at_least_one_mask_present) {
+    gdf_valid_type** masks;
+    gdf_size_type* column_lengths;
+    cudaMallocManaged((void**)&masks, sizeof(gdf_valid_type*)*num_columns);
```
Comment (jrhemstad) [inline src/column.cpp]: Can you wrap these `cudaMallocManaged` calls in `CUDA_TRY`?

File: src/column.cpp [hunk 4]
```diff
@@ -1,13 +1,171 @@
...
+    cudaMallocManaged((void**)&column_lengths, sizeof(gdf_size_type)*num_columns);
+
+    for (int i = 0; i < num_columns; ++i) {   
+      masks[i] = columns_to_concat[i]->valid;
+      column_lengths[i] = columns_to_concat[i]->size;
+    }
+  
+    gdf_mask_concat(output_column->valid, output_column->size, masks, column_lengths, num_columns);
+
+    cudaFree(masks);
```
Comment (jrhemstad) [inline src/column.cpp]: Wrap in `CUDA_TRY`

File: src/column.cpp [hunk 5]
```diff
@@ -1,13 +1,171 @@
...
+    cudaFree(column_lengths);
+  }
+  else {
+    // no masks, so just fill output valid mask with all 1 bits
+    // TODO: async
+    cudaMemset(output_column->valid, 0xff, gdf_get_num_chars_bitmask(total_size) * sizeof(gdf_valid_type));
```
Comment (jrhemstad) [inline src/column.cpp]: Wrap in `CUDA_TRY`

File: src/column.cpp [hunk 6]
```diff
@@ -1,13 +1,171 @@
...
+  //       (this will require getting rid of the allocations below because they will
+  //       implicitly sync the device), or
+  //    b) use a kernel to copy the data from all columns in one go. This will likely not
+  //       overlap with the gdf_mask_concat
+  // 2. Detect a zero total null count and skip gdf_mask_concat -- use cudaMemsetAsync
+
+  // copy data
+  for (int i = 0; i < num_columns; ++i) {   
+    gdf_size_type bytes = sizeof(column_type) * columns_to_concat[i]->size;
+    cudaMemcpy(target, columns_to_concat[i]->data, bytes, cudaMemcpyDeviceToDevice);
```
Comment (jrhemstad) [inline src/column.cpp]: Wrap in `CUDA_TRY`

File: src/column.cpp (hunk 5 above)
Comment (jrhemstad) [reply src/column.cpp]: Also, add a check to make cure `nullptr != outputcolumn->valid`

**Comment by harrism** [comment]
rerun tests

File: src/column.cpp [hunk 7]
```diff
@@ -1,13 +1,182 @@
...
+  }
+  
+  return GDF_SUCCESS;
+}
+
+/** ---------------------------------------------------------------------------*
+ * @brief Return the size of the gdf_column data type.
+ *
+ * @return gdf_size_type Size of the gdf_column data type.
+ *b  ---------------------------------------------------------------------------**/
```
Comment (scopatz) [inline src/column.cpp]: Weird that this line has a `b` in it

**Comment by scopatz** [comment]
LGTM!

**Comment by kkraus14** [comment]
cc @nsakharnykh @mt-jones

**Comment by scopatz** [comment]
There are now conflicts on master

**Comment by jrhemstad** [comment]
@dantegd can you verify that this has been sufficiently tested on the Python side?

**Pull request merged**